Kernel_Documentation/Evaluating_and_improving_the_Linux_Kernel_documentation.md

```diff
@@ -0,0 +1,113 @@
+
+# Assessment of the currently available documentation
+
+An initial assessment of the currently available documentation was performed to verify the following statement from the section “[Writing kernel-doc comments](https://www.kernel.org/doc/html/latest/doc-guide/kernel-doc.html)” in the Linux Kernel Documentation: “Every function that is exported to loadable modules using EXPORT_SYMBOL or EXPORT_SYMBOL_GPL should have a kernel-doc comment”.<BR />
+In order to have a kernel-doc header associated with a symbol come up in the official Kernel Documentation the following should happen:
+
+1. The corresponding file exporting the symbol shall be cross-referenced from the rst file in the Kernel documentation folder;
+2. The symbol shall be documented with a Kernel doc header.
+
+The initial assessment focused on point 1. and consisted in running the “[find_unreferenced_sources.sh](https://github.com/elisa-tech/Safety_Architecture_WG/blob/main/Kernel_Documentation_tools/find_unreferenced_sources.sh)” script on the Linux kernel source tree (see the output [here](./exported_symbols_files_cross_reference_report)). From the report we have 6069 source files exporting symbols and only 547 of them are cross-referenced from the Linux Kernel documentation; therefore today more than 90% of files to be cross-referenced for symbols documentation are not.<BR />
+As a final note, the initial assessment did not cover eventually duplicated information between different repos (e.g. symbol specific info already documented in the md files); this is something that will be covered later after a suitable documentation process has been defined.<BR />
+Given the significant gap already found, a [session](https://lpc.events/event/18/contributions/1894/) was held at Linux Plumbers 2024 to explain the software design aspects that are expected to be documented by quality and safety international standards, where the Linux documentation process stands in that regard and what should be done in order to fill the gaps. The agreed next steps are as follows:
+
+1. Defining a template (or templates) that can be used to document Kernel code / User APIs: sysfs APIs have been proposed as a good starting point to elaborate on the aspects that can be covered by such template;
+2. Getting such a template (or templates) reviewed internally in ELISA first and then externally by a restricted group of maintainers (Jonathan Corbet, Thomas Gleixner, Steve Rostedt, Bjorn Andersson);
+3. Starting to document a specific subsystem using such templates (possibly mm) and getting the documentation upstream;
+4. Having LF to kick-off a new project for extending such documentation and for defining the process to enforce it to the rest of the Kernel.<BR /><BR />
```

Comment:
There was the explicit request that we develop a tool which can be incorporated in the Kernel CI/CD to verify if the documentation is done accordingly. Do you want to mention it here, or did you explicitly choose to formulate it more generically in this 4. aspect?